### DISCUSSION TITLE
Add support for Algerian weather alerts

### DISCUSSION THREAD

**Discussion by naitgacem** [breezy-weather/breezy-weather#1283] [Ideas]
### Describe your idea

The algerian National Meteorological Office offers an [RSS feed](https://ametvigilance.meteo.dz/rss/rss_meteo_dz.xml) for weather alerts. Perhaps we can incorporate this into the app? The Feed contains `<polygon>`  coordinates that can be parsed as an area on the map. It broadcasts weather condition alerts: events such as `Rain`, `Storm`, ... as well as the severity and certainty information.
If the app already supports multiple weather alerts sources, perhaps we can extend it with an Algerian specific one.

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.7 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by papjul** [comment]
We already have support for Météo Algérie RSS through the WMO Severe Weather project. Unfortunately, since we don't have a XML parser at the moment (see #808), the description may be truncated since we can't parse the RSS (yet), but only what WMO makes available as JSON.

**Comment by naitgacem** [comment]
If I understand this correctly, the blocking issue here is the CAP parser? Can you point me to the documentation of the WMO API?
I see in the code that the requests are being sent to `severeweather.wmo.int/f/wfs` But I can't find any documentation.

**Reply by papjul** [reply]
It's a WFS service, so you can get the documentation by making the queries you want (such as `GetCapabilities`).

However, all you need to know is that in our case, it returns a JSON with for each alert a `capUrl` or `url` attribute depending on the national service providing it.

You need to prefix it:
- For `capUrl`: `https://8xieiqdnye.execute-api.us-west-2.amazonaws.com/swic/capUrl/`
- For `url`: `https://cvzxdcwxid.execute-api.us-west-2.amazonaws.com/swic/url/`

Then when accessing it, you will get a XML file which we don't parse yet (we only use JSON API so far).

**Reply by naitgacem** [reply]
Okay so i made a request and here is the resulting JSON (minus the polygon part since it's large).
```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "id": "postgis_geojsons.fid--183103f2_191b6f0ec7e_3464",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          ...
        ]
      },
      "geometry_name": "wkb_geometry",
      "properties": {
        "capurl": "dz-onm-fr/2024/09/03/06/06/52-89c38d445a5bfe0ef0aa671061a14cc2.xml",
        "identifier": null,
        "sent": "2024-09-02T22:06:52Z",
        "description": "No description",
        "event": "ORAGE",
        "s": 2,
        "u": 3,
        "c": 3,
        "sn": "",
        "effective": "2024-09-02T22:06:52Z",
        "onset": "2024-09-03T13:00:00Z",
        "mem": "122",
        "marine": "0",
        "areadesc": "NAÂMA",
        "url": "2024-09-03 14:15:03",
        "expires": "2024-09-03T16:00:00Z",
        "gc": "",
        "cn": null,
        "headline": "",
        "chk_effective": "2024-09-03T13:00:00Z",
        "chk_expires": "2024-09-03T16:00:00Z",
        "rlink": "",
        "file_name": "algeria.geojson",
        "row_type": "POLYGON"
      },
      "bbox": [
        -1.73708994,
        32.14227362,
        0.15829006,
        34.32262833
      ]
    }
  ],
  "totalFeatures": 1,
  "numberMatched": 1,
  "numberReturned": 1,
  "timeStamp": "2024-09-03T08:13:56.407Z",
  "crs": {
    "type": "name",
    "properties": {
      "name": "urn:ogc:def:crs:EPSG::4326"
    }
  },
  "bbox": [
    32.14227362,
    -1.73708994,
    34.32262833,
    0.15829006
  ]
}
```
Now the URL is supposed to be 
`https://8xieiqdnye.execute-api.us-west-2.amazonaws.com/swic/capUrl/dz-onm-fr/2024/09/03/06/06/52-89c38d445a5bfe0ef0aa671061a14cc2.xml`

However the response is `No such key`. So perhaps there is somwthing missing?

**Reply by papjul** [reply]
Looks like the prefix URL changed since I last checked. It is now: `https://severeweather.wmo.int/v2/cap-alerts/`

**Reply by naitgacem** [reply]
Got it. As for parsing XML, is somwthing like Java's DOM parser not enough? Has there been any discussion about this?

**Reply by papjul** [reply]
We currently use Kotlin Serialization.

There is a XML plugin for it but there is no stable release yet, making it impossible to make reproducible builds: https://github.com/rharter/kotlinx-serialization-xml

I have not yet studied unofficial options such as https://github.com/pdvrieze/xmlutil

**Reply by papjul** [reply]
In your case, I don't think you will need to parse anything. From what I can see, your national source only provides data in French so there won't be additional translations in the XML. The descriptions are short, so they are not cut. Tl;dr: the WMO source has everything you need

**Comment by papjul** [comment]
The RSS feed cannot be used as it contains the alerts from the whole country.
Knowing if an alert is emitted for the location would require opening each alert URL one by one and check the polygon.
Suggested replacement is WMO SWIC which use the same data, but proceed the data first so we only get the alerts for the location, and not for the whole country.
Closing as “Not feasible”.
